tools/depmod.c

```diff
@@ -2610,53 +2619,42 @@ static int depmod_output(struct depmod *depmod, FILE *out)
 		int r, ferr;
 
 		if (fp == NULL) {
-			int flags = O_CREAT | O_EXCL | O_WRONLY;
-			int mode = 0644;
-			int fd;
-			int n;
-
-			n = snprintf(tmp, sizeof(tmp), "%s.%i.%lli.%lli", itr->name,
-				     getpid(), (long long)tv.tv_usec,
-				     (long long)tv.tv_sec);
-			if (n >= (int)sizeof(tmp)) {
-				ERR("bad filename: %s.%i.%lli.%lli: path too long\n",
-				    itr->name, getpid(), (long long)tv.tv_usec,
-				    (long long)tv.tv_sec);
+			size_t len = strlen(dpath) + strlen(itr->name) + 2;
+			if (len >= PATH_MAX) {
+				ERR("bad filename: %s/%s: path too long\n", dpath,
+				    itr->name);
 				continue;
 			}
-			fd = openat(dfd, tmp, flags, mode);
-			if (fd < 0) {
-				ERR("openat(%s, %s, %o, %o): %m\n", dname, tmp, flags,
-				    mode);
-				continue;
-			}
-			fp = fdopen(fd, "wb");
-			if (fp == NULL) {
-				ERR("fdopen(%d=%s/%s): %m\n", fd, dname, tmp);
-				close(fd);
+			memset(outputname, 0, PATH_MAX);
+			snprintf(outputname, len, "%s/%s", dpath, itr->name);
+
+			err = tmpfile_init(&file, outputname);
+			if (err < 0) {
+				ERR("fail to create temporary for (%s): %m\n", outputname);
+				tmpfile_release(&file);
 				continue;
 			}
+			fp = file.f;
 		}
 
 		r = itr->cb(depmod, fp);
 		if (fp == out)
 			continue;
 
-		ferr = ferror(fp) | fclose(fp);
+		ferr = ferror(fp);
 
 		if (r < 0) {
-			if (unlinkat(dfd, tmp, 0) != 0)
-				ERR("unlinkat(%s, %s): %m\n", dname, tmp);
+			tmpfile_release(&file);
 
 			ERR("Could not write index '%s': %s\n", itr->name, strerror(-r));
 			err = -errno;
 			break;
 		}
 
-		if (renameat(dfd, tmp, dfd, itr->name) != 0) {
-			err = -errno;
-			CRIT("renameat(%s, %s, %s, %s): %m\n", dname, tmp, dname,
-			     itr->name);
+		err = tmpfile_publish(&file);
+		if (err != 0) {
+			CRIT("publish temporary from %s to %s: %m\n", file.tmpname,
```

Comment:
I don't thinks we can use %m anymore. If the function returned -EINVAL, that's what the should tell the user, not a possible old errno set by libc (note that in your implementation there isn't even a libc call at that point, so it would probably read "SUCCESS" in the error message.